controller_test.go

```diff
@@ -0,0 +1,141 @@
+	nodeName string
+}
+
+func (r *runtimeClientImpt) Get(ctx context.Context, key types.NamespacedName, obj client.Object, opts ...client.GetOption) error {
+	obj.SetLabels(r.labels)
+	obj.SetName(r.nodeName)
+	return nil
+}
+
+type instanceClientImpl struct {
+	calledWithProjectID        string
+	calledWithZone             string
+	calledWithInstanceName     string
+	calledWithSecurityPolicy   string
+	calledWithNetworkInterface []string
+	timesCalled                int
+}
+
+func (i *instanceClientImpl) SetSecurityPolicy(ctx context.Context, req *computepb.SetSecurityPolicyInstanceRequest, opts ...gax.CallOption) (*compute.Operation, error) {
+	i.calledWithProjectID = req.Project
+	i.calledWithZone = req.Zone
+	i.calledWithInstanceName = req.Instance
+	i.calledWithSecurityPolicy = *req.InstancesSetSecurityPolicyRequestResource.SecurityPolicy
+	i.calledWithNetworkInterface = req.InstancesSetSecurityPolicyRequestResource.NetworkInterfaces
+	i.timesCalled++
+	return nil, nil
+}
+
+func TestReconcileTable(t *testing.T) {
+	var tests = []struct {
+		name              string
+		label             string
+		value             string
+		zone              string
+		nodeName          string
+		selector          string
+		securityPolicyURL string
+		projectID         string
+		timesCalled       int
+	}{
+		{
+			"label is matched",
+			"cloud.google.com/gke-nodepool",
+			"default-pool",
+			"us-central1-a",
+			"my-node-name",
+			"cloud.google.com/gke-nodepool=default-pool",
+			"securityPolicyURL",
+			"my-project-id",
+			1,
+		},
+		{
+			"label is not matched",
+			"cloud.google.com/gke-nodepool",
+			"default-pool",
+			"us-central1-a",
+			"my-node-name",
+			"cloud.google.com/gke-nodepool=not-default-pool",
+			"securityPolicyURL",
+			"my-project-id",
+			0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			nodeLabels := make(map[string]string)
+			nodeLabels[tt.label] = tt.value
+			nodeLabels["topology.gke.io/zone"] = tt.zone
+			name := tt.nodeName
+			parsedSelector, err := labels.Parse(tt.selector)
+			if err != nil {
+				t.Errorf("error parsing selector %v, want no error", err)
+			}
+			rc := runtimeClientImpt{labels: nodeLabels, nodeName: name}
+			ic := instanceClientImpl{}
+
+			r := &reconcileNode{
+				client:            &rc,
+				instanceClient:    &ic,
+				selector:          parsedSelector,
+				securityPolicyURL: tt.securityPolicyURL,
+				projectID:         tt.projectID,
+			}
+
+			reconcileReq := reconcile.Request{}
+
+			if _, err := r.Reconcile(context.Background(), reconcileReq); err != nil {
+				t.Errorf("error %v, want no error", err)
+			}
+
+			if ic.timesCalled != tt.timesCalled {
+				t.Errorf("wrong number of times called: %d, want %d", ic.timesCalled, tt.timesCalled)
+			}
+
+			if ic.timesCalled == 0 {
+				// Skip checking the SetSecurityPolicyInstanceRequest because there is none.
+				return
+			}
+
+			if ic.calledWithSecurityPolicy != tt.securityPolicyURL {
```

Comment:
optional nit: this field by field check is ok, but I'd say a common approach is also just to store a reference to the structure and use cmp.Diff against the whole structure: https://github.com/google/go-cmp/blob/master/cmp/example_test.go#L30-L32

Reply:
I tried it but got 2 problems:

first is a warning saying "call of cmp.Diff copies lock value: cloud.google.com/go/compute/apiv1/computepb.SetSecurityPolicyInstanceRequest contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex"

if I ignore the warning and run the test, the test fails with 

=================================================
-- FAIL: TestReconcileTable (0.00s)
    --- FAIL: TestReconcileTable/label_is_matched (0.00s)
panic: cannot handle unexported field at {computepb.SetSecurityPolicyInstanceRequest}.state:
        "cloud.google.com/go/compute/apiv1/computepb".SetSecurityPolicyInstanceRequest
consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported [recovered]
        panic: cannot handle unexported field at {computepb.SetSecurityPolicyInstanceRequest}.state:
        "cloud.google.com/go/compute/apiv1/computepb".SetSecurityPolicyInstanceRequest
=================================================

Seems like this request can not easily be compared?

Reply:
Oh, just use `cmpopts.IgnoreUnexported`. You don't need to compare the unexported values.

Reply:
Oh done